man/gsm_option.3
.\"
.\" Copyright 1992-1995 by Jutta Degener and Carsten Bormann, Technische
.\" Universitaet Berlin.  See the accompanying file "LICENSE" for
.\" details.  THERE IS ABSOLUTELY NO WARRANTY FOR THIS SOFTWARE.
.\"
.PU
.TH GSM_OPTION 3 
.SH NAME
gsm_option \(em customizing the GSM 06.10 implementation
.SH SYNOPSIS
#include "gsm.h"
.PP
int gsm_option(handle, option, valueP);
.br
gsm handle;
.br
int option;
.br
int * valueP;
.SH "DESCRIPTION"
The gsm library is an implementation of the final draft GSM 06.10
standard for full-rate speech transcoding, a lossy
speech compression algorithm.
.PP
The gsm_option() function can be used to set and query various
options or flags that are not needed for regular GSM 06.10 encoding
or decoding, but might be of interest in special cases.
.PP
The second argument to gsm_option specifies what parameter
should be changed or queried.
The third argument is either a null pointer, in which case
the current value of that parameter is returned;
or it is a pointer to an integer containing the value
you want to set, in which case the previous value will
be returned.
.PP
The following options are defined:
.PP
.I GSM_OPT_VERBOSE
Verbosity level.
.br
.in+5
This option is only supported if the library was compiled
with debugging turned on, and may be used by developers of
compression algorithms to aid debugging.
.br
The verbosity level can be changed at any time during encoding or decoding.
.in-5
.sp
.PP
.I GSM_OPT_FAST
Faster compression algorithm.
.br
.in+5
This implementation offers a not strictly standard-compliant, but
faster compression algorithm that is compatible with the regular
method and does not noticably degrade audio quality.
.br
The value passed to 
.br
.nf
	gsm_option(handle, GSM_OPT_FAST, & value)
.fi
.br 
functions as a boolean flag; if it is zero, the regular algorithm
will be used, if not, the faster version will be used.
.br
The availability of this option depends on the hardware used;
if it is not available, gsm_option will return -1 on an attempt
to set or query it.
.br
This option can be set any time during encoding or decoding.
.in-5
.ne 5
.sp
.PP
.I GSM_OPT_LTP_CUT
Enable, disable, or query the LTP cut-off optimization.
.br
.in+5
During encoding, the search for the long-term correlation
lag forms the bottleneck of the algorithm. 
The ltp-cut option enables an approximation that disregards most
of the samples for purposes of finding that correlation,
and hence speeds up the encoding at a noticable loss in quality.
.br
The value passed to 
.br
.nf
	gsm_option(handle, GSM_OPT_LTP_CUT, & value)
.fi
.br 
turns the optimization on if nonzero, and off if zero.
.br
This option can be set any time during encoding
or decoding; it will only affect the encoding pass, not
the decoding.
.sp
.PP
.I GSM_OPT_WAV49
WAV-style byte ordering.
.br
.in+5
A WAV file of type #49 contains GSM 06.10-encoded frames.
Unfortunately, the framing and code ordering of the WAV version
are incompatible with the native ones of this GSM 06.10 library.
The GSM_OPT_WAV49 option turns on a different packing
algorithm that produces alternating frames of 32 and 33 bytes
(or makes it consume alternating frames of 33 and 32 bytes, note
the opposite order of the two numbers) which, when concatenated,
can be used in the body of a WAV #49 frame.
It is up to the user program to write a WAV header, if any;
neither the library itself nor the toast program produce
complete WAV files.
.br
The value passed to 
.br
.nf
	gsm_option(handle, GSM_OPT_WAV49, & value)
.fi
.br 
functions as a boolean flag; if it is zero, the library's native
framing algorithm will be used, if nonzero, WAV-type packing is in effect.
.br
This option should be used before any frames are encoded.
Whether or not it is supported at all depends on a
compile-time switch, WAV49.
Both option and compile time switch are new to the library
as of patchlevel 9, and are considerably less tested than the
well-worn rest of the it.
.br
Thanks to Jeff Chilton for the detective work and first free
implementation of this version of the GSM 06.10 encoding.
.sp
.PP
.I GSM_OPT_FRAME_CHAIN
Query or set the chaining byte.
.br
.in+5
Between the two frames of a WAV-style encoding, the GSM 06.10 library
must keep track of one half-byte that is technically part of the first
frame, but will be written as the first four bits of the second.
This half-byte are the lowest four bits of the value returned by,
and optionally set by,
.br
.nf
	gsm_option(handle, GSM_OPT_FRAME_CHAIN, & value)
.fi
.br 
This option can be queried and set at any time.
.sp
.PP
.I GSM_OPT_FRAME_INDEX
Query or set the current frame's index in a format's
alternating list of frames.
.br
.in+5
The WAV #49 framing uses two alternating types of frames.
Which type the next GSM-coded frame belongs to can be queried, or,
when decoding, announced, using
.br
.nf
	gsm_option(handle, GSM_OPT_FRAME_INDEX, & value)
.fi
.br 
For WAV-style framing, the value should be 0 or 1; the first frame
of an encoding has an index of 0. 
At library initialization, the index is set to zero.
.br 
The frame index can be queried and set at any time.
Used in combination with the
.IR GSM_OPT_FRAME_CHAIN ,
option, it can be used to position on arbitrary GSM frames
within a format like WAV #49 (not accounting for the lost
internal GSM state).
.in-5
.SH "RETURN VALUE"
gsm_option() returns -1 if an option is not supported, the
previous value of the option otherwise.
.SH BUGS
Please direct bug reports to jutta@pobox.com and cabo@tzi.org.
.SH "SEE ALSO"
toast(1), gsm(3), gsm_explode(3), gsm_print(3)


ChangeLog
Thu Aug  4 18:47:10 2022	Jutta Degener (jutta@pobox.com)

	* Release 1.0 Patchlevel 22
	Remove unused, and occasionally out of date, VERSION variable
	from Makefile.

Sun Jul 31 16:17:59 2022	Jutta Degener (jutta@pobox.com)

	* Release 1.0 Patchlevel 21
	Use $(LN), not ln, in the Makefile rules for the installed binary
 	of untoast and tcat.
	(Thanks to Carlo Cabrera for pointing that out!)

Mon Jul  4 10:34:18 2022

	* Release 1.0 Patchlevel 20
	Use $(RMFLAGS) with the rms in the Makefile; add -f to $(RMFLAGS) to
        avoid spurious error messages during build.

Fri Mar 27 15:18:06  2020

	* Release 1.0 Patchlevel 19
	Make it easier to include gsm.h from C++ by wrapping it with macros. 
	Throw a #define _POSIX_C_SOURCE 200809L into toast.h to make it
	possible for Linux glibc to just work.
	(Thanks to enh@google.com for the suggestion; errors are mine.)
	Switch from compress/.Z to gzip/.gz 

Sat May  5 21:48:20  2018

	* Release 1.0 Patchlevel 18
	Remove the MacOS "quarantine" files that I didn't know I'd shipped.
	Remove some trailing spaces.
	Bump up the patchlevel in inc/gsm.h (I'd forgotten that in release 17.)

Tue Jun 27 17:59:30  2017

	* Release 1.0 Patchlevel 17
	New email addresses in the README.
	Avoid left-shifting signed integers.
	Don't print longwords using %d or %ld, cast them to (long) first.
	Don't compare longwords as longs in addtest, we only care for the
		lower 32 bits. (The math works, but the test failed.)

Sat Sep 17 14:50:55  2016

	* Release 1.0 Patchlevel 16
	New email addresses in the manual pages (patch 16)
	Removed a stray .orig file (in patch 15) 

Sun Apr  5  04:53:07 2009       Jutta Degener (jutta@pobox.com)

	* Release 1.0 Patchlevel 13
	COPYRIGHT: exploring the deeper meaning and cultural
	history of the word "any".

Fri Jun  2 06:27:56 2006	Jutta Degener (jutta@pobox.com)

	* Release 1.0 Patchlevel 12
	src/code.c: removed modifiable static storage that made the library
	 	non-reentrant.  (Thanks to Paul Gibbs for pointing out the bug.)

Sat Mar 18 11:48:28 2006	Jutta Degener (jutta@pobox.com)

	* Release 1.0 Patchlevel 11
	src/gsm_implode.c: fix WAV49 implode
	move source archive location

Fri Jul  5 19:26:37 1996 	Jutta Degener (jutta@cs.tu-berlin.de)

	* Release 1.0 Patchlevel 10
	src/toast_alaw.c: exchanged A-law tables for something
		slightly more A-law.

Tue Jul  2 12:18:20 1996  Jutta Degener (jutta@cs.tu-berlin.de)

	* Release 1.0 Patchlevel 9
	src/long_term.c: in FLOAT_MUL mode, an array was accessed past its end
	src/gsm_option.c: three options related to WAV #49 packing
	src/gsm_encode.c: support WAV #49-style encoding.
	src/gsm_decode.c: support WAV #49-style decoding.
	tls/sour.c: generate the WAV bit shifting code, encode
	tls/ginger.c: generate the WAV bit shifting code, decode
	The WAV code goes back to an inofficial patch #8 that
	Jeff Chilton sent us (hence the jump from 7 to 9).
	src/toast.c: add _fsetmode() calls to set stdin/stdout to
		binary (from an OS/2 port by Arnd Gronenberg.)

Tue Mar  7 01:55:10 1995  Jutta Degener (jutta@cs.tu-berlin.de)

	* Release 1.0 Patchlevel 7
	src/long_term.c: Yet another 16-bit overflow
	src/toast.c: -C option to toast, cuts LPC time
	src/gsm_option.c: corresponding LPC_CUT option to GSM library

Fri Dec 30 23:33:50 1994  Jutta Degener (jutta@cs.tu-berlin.de)

        * Release 1.0 Patchlevel 6
        src/lpc.c: fixed 16-bit addition overflow in Autocorrelation code
        src/add.c: gsm_L_asl should fall back on gsm_L_asr, not gsm_asr

Mon Nov 28 20:49:57 1994  Jutta Degener (jutta@cs.tu-berlin.de)
	
	* Release 1.0 Patchlevel 5
	src/toast_audio.c: initialization should return -1 on error
	src/gsm_destroy.c: #include configuration header file
	src/add.c: gsm_sub should cast its parameters to longword
	man/*: bug reports to {jutta,cabo}@cs.tu-berlin.de, not to toast@tub
	inc/private.h: longword long by default, not int
	inc/toast.h: read/write fopen modes "rb" and "wb", not just "r"
	src/toast.c: better (or different, anyway) error handling in process()

Tue May 10 19:41:34 1994  Jutta Degener (jutta at kugelbus)
	
	* Release 1.0 Patchlevel 4
	inc/private.h: GSM_ADD should cast to ulongword, not to unsigned.
	src/long_term.c: missing cast to longword.
	add-test/add_test.c: Test macros too, not only functions,
	thanks to Simao Ferraz de Campos Neto, simao@dragon.cpqd.ansp.br
	General cleanup: remove unused variables, add function prototypes.

Tue Jan 25 22:53:40 1994  Jutta Degener (jutta at kugelbus)

	* Release 1.0 Patchlevel 3
	changed rpe.c's STEP macro to work with 16-bit integers,
	thanks to Dr Alex Lee (alexlee@solomon.technet.sg);
	removed non-fatal bugs from add-test.dta, private.h
	and toast_audio.c, thanks to P. Emanuelsson.

Fri Jan 29 19:02:12 1993  Jutta Degener  (jutta at kraftbus)
	
	* Release 1.0 Patchlevel 2
	fixed L_add(0,-1) in src/add.c and inc/private.h,
	thanks to Raphael Trommer at AT&T Bell Laboratories;
	various other ANSI C compatibility details

Fri Oct 30 17:58:54 1992  Jutta Degener  (jutta at kraftbus)

	* Release 1.0 Patchlevel 1
	Switched uid/gid in toast's [f]chown calls.

Wed Oct 28 14:12:35 1992  Carsten Bormann  (cabo at kubus)

	* Release 1.0: released
	Copyright 1992 by Jutta Degener and Carsten Bormann, Technische
	Universitaet Berlin.  See the accompanying file "LICENSE" for
	details.  THERE IS ABSOLUTELY NO WARRANTY FOR THIS SOFTWARE.



src/gsm_option.c
/*
 * Copyright 1992 by Jutta Degener and Carsten Bormann, Technische
 * Universitaet Berlin.  See the accompanying file "LICENSE" for
 * details.  THERE IS ABSOLUTELY NO WARRANTY FOR THIS SOFTWARE.
 */

/* $Header: /tmp_amd/presto/export/kbs/jutta/src/gsm/RCS/gsm_option.c,v 1.3 1996/07/02 09:59:05 jutta Exp $ */

#include "private.h"

#include "gsm.h"
#include "proto.h"

int gsm_option P3((r, opt, val), gsm r, int opt, int * val)
{
	int 	result = -1;

	switch (opt) {
	case GSM_OPT_LTP_CUT:
#ifdef 	LTP_CUT
		result = r->ltp_cut;
		if (val) r->ltp_cut = *val;
#endif
		break;

	case GSM_OPT_VERBOSE:
#ifndef	NDEBUG
		result = r->verbose;
		if (val) r->verbose = *val;
#endif
		break;

	case GSM_OPT_FAST:

#if	defined(FAST) && defined(USE_FLOAT_MUL)
		result = r->fast;
		if (val) r->fast = !!*val;
#endif
		break;

	case GSM_OPT_FRAME_CHAIN:

#ifdef WAV49
		result = r->frame_chain;
		if (val) r->frame_chain = *val;
#endif
		break;

	case GSM_OPT_FRAME_INDEX:

#ifdef WAV49
		result = r->frame_index;
		if (val) r->frame_index = *val;
#endif
		break;

	case GSM_OPT_WAV49:

#ifdef WAV49
		result = r->wav_fmt;
		if (val) r->wav_fmt = !!*val;
#endif
		break;

	default:
		break;
	}
	return result;
}
